common/achievement/index.ts

```diff
@@ -0,0 +1,123 @@
+import { prisma } from '../prisma';
+import { User } from '../user';
+import { isGamificationFeatureActive, getMetricsByAction, sortActionTemplatesToGroups } from './util';
+import { getLogger } from '../logger/logger';
+import { ActionID, SpecificNotificationContext } from '../notification/actions';
+import { getTemplatesByAction } from './template';
+import { evaluateAchievement } from './evaluate';
+import { AchievementToCheck, ActionEvent, ConditionDataAggregations, UserAchievementContext, UserAchievementTemplate } from './types';
+import { createAchievement, getOrCreateUserAchievement } from './create';
+import { actionTakenAt } from '../notification';
+
+const logger = getLogger('Achievement');
+
+export async function rewardActionTaken<ID extends ActionID>(user: User, actionId: ID, context: SpecificNotificationContext<ID>) {
+    if (!isGamificationFeatureActive()) {
+        return;
+    }
+    const templatesForAction = await getTemplatesByAction(actionId);
+
+    if (templatesForAction.length === 0) {
+        logger.debug(`No achievement found for action '${actionId}'`);
+        return;
+    }
+
+    const templatesByGroups = sortActionTemplatesToGroups(templatesForAction);
+
+    const actionEvent: ActionEvent<ID> = {
+        actionId,
+        at: new Date(),
+        user: user,
+        context,
+    };
+    await trackEvent(actionEvent);
+
+    for (const [, group] of templatesByGroups) {
+        let achievementToCheck: AchievementToCheck;
+        for (const template of group) {
+            const userAchievement = await getOrCreateUserAchievement(template, user.userID, context);
+            if (userAchievement.achievedAt === null || userAchievement.recordValue) {
```

Comment:
I think `userAchievement` can be undefined if it's the last in group.

Reply:
why could a userAchievement be undefined. Can not follow.